### PR TITLE
Add Diskuv.OCaml.1.0.0.0

### PR DESCRIPTION
**Pull request by jonahbeckford**
- [x] Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/microsoft/winget-pkgs/pulls) for the same manifest update/change?
- [x] Have you [validated](https://github.com/microsoft/winget-pkgs/blob/master/AUTHORING_MANIFESTS.md#validation) your manifest locally with `winget validate --manifest <path>`? 
- [x] Have you tested your manifest locally with `winget install --manifest <path>`?
- [x] Does your manifest conform to the [1.1 schema](https://github.com/microsoft/winget-pkgs/tree/master/doc/manifest/schema/1.1.0)?

Note: `<path>` is the name of the directory containing the manifest you're submitting.

-----

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-pkgs/pull/69112)

### CODE REVIEW

File: manifests/d/Diskuv/OCaml/1.0.0.0/Diskuv.OCaml.installer.yaml [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+      SilentWithProgress: --ci
+      Interactive: --color=always
+      InstallLocation: --prefix "<INSTALLPATH>"
+
+    AppsAndFeaturesEntries:
+      # Keep in sync with installer\src\private_common.ml which is used by
+      # https://github.com/diskuv/dkml-install-api/blob/main/package/console/common/Windows_registry.ml
+      - Publisher: Diskuv, Inc.
+        DisplayName: Diskuv OCaml
+        DisplayVersion: 1.0.0.0
```
Comment (Trenly) [inline manifests/d/Diskuv/OCaml/1.0.0.0/Diskuv.OCaml.installer.yaml:68]:
```suggestion
        DisplayVersion: dev
```
This must exactly match what is written into control panel

**Review by Trenly — CHANGES_REQUESTED** [review]

**Review by ItzLevvie — DISMISSED** [review]

**Comment by jonahbeckford** [comment]
Oops. I investigated and forgot to update this ticket. Let me paste what I did and my findings ...

**Comment by jonahbeckford** [comment]
> You can test with https://github.com/microsoft/winget-pkgs#test-your-manifest.

> ... You may also try using the [Windows Package Manager Manifest Creator](https://github.com/microsoft/winget-create)

I used Windows Package Manager Manifest Creator, and then run it inside a Windows Sandbox to see what was happening. I also adjusted several fields. But I consistently ran into:

Sandbox-2022-08-12-101842

**Comment by jonahbeckford** [comment]
But if I download and run my installer directly in the Windows Sandbox, it works:

Sandbox-2022-08-12 102439-DuringInstall

**Comment by jonahbeckford** [comment]
My installer is a 7zip self-extracting executable. The sequence is (applies to all 7zip self-extracting executables):

1. Decompress the setup.exe into a temporary directory
2. Launch an executable that is in the root folder of the temporary directory

I have "silent" switches to remove user interaction, but it is not even getting that far. Instead, it immediately after download (before step #1) says "Search for app in the Store".

Do you know what the problem is? Or at least some pointers to let me investigate further?

Thanks!

**Comment by jonahbeckford** [comment]
FYI: I have comments about the failure above.

**Comment by ItzLevvie** [comment]
> Instead, it immediately after download (before step 1) says "Search for app in the Store".

It's because "Microsoft Defender SmartScreen" (the blue prompt) has been removed from Windows Sandbox in Windows 11.

You can workaround it by running this in Windows PowerShell to disable Microsoft Defender SmartScreen:
```powershell
if ($env:USERNAME -eq "WDAGUtilityAccount") {
    if (-not(Get-ItemProperty -Path HKLM:\SOFTWARE\Microsoft\Windows\CurrentVersion\Explorer -Name SmartScreenEnabled)) {
        Set-ItemProperty -Path HKLM:\SOFTWARE\Microsoft\Windows\CurrentVersion\Explorer -Name SmartScreenEnabled -Value Off -Force
    }
}
```

> Do you know what the problem is? Or at least some pointers to let me investigate further?

When I tried to install it - it installed many dependencies and started to compile https://github.com/ocaml/ocaml:
image
image

I started the install at 8:43 PM and it finished at 10:42 PM which is roughly 2 hours (depending on the system configuration used) to install:
```
Enter a pull request number: 69112

Manifest validation succeeded.
Found Diskuv OCaml [Diskuv.OCaml] Version 1.0.0.0
This application is licensed to you by its owner.
Microsoft is not responsible for, nor does it grant any licenses to, third-party packages.
Downloading https://github.com/diskuv/dkml-installer-ocaml/releases/download/v1.0.0/setup-diskuv-ocaml-windows_x86_64-1.0.0.exe
  ██████████████████████████████  87.2 MB / 87.2 MB
Successfully verified installer hash
Starting package install...
Successfully installed

Name          : Diskuv OCaml
Publisher     : Diskuv, Inc.
Version       : dev
ProductCode   : DiskuvOCaml
Registry Path : Computer\HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Uninstall\DiskuvOCaml
Uninstall     : winget uninstall --id "DiskuvOCaml"

Name          : Microsoft Visual C++ 2015-2019 Redistributable (x64) - 14.29.30133
Publisher     : Microsoft Corporation
Version       : 14.29.30133.0
ProductCode   : {295d1583-fdb9-414b-a4c8-da539362a26b}
Registry Path : Computer\HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\Uninstall\{295d1583-fdb9-414b-a4c8-da539362a26b}
Uninstall     : winget uninstall --id "{295d1583-fdb9-414b-a4c8-da539362a26b}"

Name          : Microsoft Visual C++ 2015-2019 Redistributable (x86) - 14.29.30133
Publisher     : Microsoft Corporation
Version       : 14.29.30133.0
ProductCode   : {38b2c744-ad08-4d5b-91a2-3fb6f739ff3e}
Registry Path : Computer\HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\Uninstall\{38b2c744-ad08-4d5b-91a2-3fb6f739ff3e}
Uninstall     : winget uninstall --id "{38b2c744-ad08-4d5b-91a2-3fb6f739ff3e}"

Name          : Microsoft Visual Studio Installer
Publisher     : Microsoft Corporation
Version       : 2.11.35.61819
ProductCode   : {6F320B93-EE3C-4826-85E0-ADF79F8D4C61}
Registry Path : Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{6F320B93-EE3C-4826-85E0-ADF79F8D4C61}
Uninstall     : winget uninstall --id "{6F320B93-EE3C-4826-85E0-ADF79F8D4C61}"

Name          : Visual Studio Build Tools 2019
Publisher     : Microsoft Corporation
Version       : 16.11.2
ProductCode   : 58c31de6
Registry Path : Computer\HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\Uninstall\58c31de6
Uninstall     : winget uninstall --id "58c31de6"

Name          : Windows SDK AddOn
Publisher     : Microsoft Corporation
Version       : 10.1.0.0
ProductCode   : {E6F877A1-2F65-4BF0-87B6-A4071B7663D3}
Registry Path : Computer\HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\Uninstall\{E6F877A1-2F65-4BF0-87B6-A4071B7663D3}
Uninstall     : winget uninstall --id "{E6F877A1-2F65-4BF0-87B6-A4071B7663D3}"

Name          : Windows Software Development Kit - Windows 10.0.18362.1
Publisher     : Microsoft Corporation
Version       : 10.1.18362.1
ProductCode   : {126dedf0-cc0e-4b48-9ece-806b0e437195}
Registry Path : Computer\HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\Uninstall\{126dedf0-cc0e-4b48-9ece-806b0e437195}
Uninstall     : winget uninstall --id "{126dedf0-cc0e-4b48-9ece-806b0e437195}"

Press any key to continue . . .
```

Here you can see where Azure Pipelines is stalling:
InstallationVerification-x64-Scope_Undefined-Locale_Undefined-ErrorScreenshot_20220812181832

**Comment by jonahbeckford** [comment]
Thanks; I can troubleshoot now.

FYI for the next person with Windows 11 problems:

```powershell
# $env:USERNAME = "SYSTEM" on Windows 11 Pro 22H2, 22621.382, Windows Feature Experience Pack 1000.22632.1000.0
if (($env:USERNAME -eq "WDAGUtilityAccount") -or ($PWD.Path -eq "C:\Users\WDAGUtilityAccount\Desktop\winget-pkgs")) {
    foreach ($drive in @("HKLM", "HKCU")) {
        $path = "SOFTWARE\Microsoft\Windows\CurrentVersion\Explorer"
        $key = Get-Item -LiteralPath ${drive}:\${path}
        if (($key.GetValue("SmartScreenEnabled", $null) -eq $null) -or
            -not(Get-ItemProperty -Path ${drive}:\${path} -Name SmartScreenEnabled))
        {
            Set-ItemProperty -Path ${drive}:\${path} -Name SmartScreenEnabled -Value Off -Force
            Write-Host "${drive}:\${path} SmartScreenEnabled=Off"
        }
    }
}

# Rerun the installer; if you are prompted for security now you'll be able to click through it
winget install --manifest ..\SandboxTest\manifest
```

**Comment by jonahbeckford** [comment]
Hmm ... it does work for me in the Windows Sandbox using winget:

```
>> # Rerun the installer; if you are prompted for security now you'll be able to click through it
>> winget install --manifest ..\SandboxTest\manifest
>>
HKLM:\SOFTWARE\Microsoft\Windows\CurrentVersion\Explorer SmartScreenEnabled=Off
HKCU:\SOFTWARE\Microsoft\Windows\CurrentVersion\Explorer SmartScreenEnabled=Off
Found Diskuv OCaml [Diskuv.OCaml] Version 1.0.0.0
This application is licensed to you by its owner.
Microsoft is not responsible for, nor does it grant any licenses to, third-party packages.
Successfully verified installer hash
Starting package install...
Successfully installed
PS C:\Users\WDAGUtilityAccount\Desktop\winget-pkgs>
```

So am rerunning with updated `.yaml` to see if something I edited is making it pass for me locally.

It has things I don't like:
1. It takes ~2 hours to complete. That won't reduce until OCaml (the programming language) stops requiring the installation path to be embedded in its binaries. And that won't change for ~1 year.
2. winget does not install the dependencies Visual Studio Build Tools and Git. Nothing I place in `Dependencies:` seems to matter. It is not terrible since my installer will not install Visual Studio Build Tools and Git if they are already present and compatible, but I'm hoping to model these dependencies correctly.

And the wingetcreate is using `ManifestVersion: 1.2.0` which is too new. I'll downgrade and resubmit.

**Comment by jonahbeckford** [comment]
This looks like a 2hr time limit since this succeeds in Windows Sandbox. I can withdraw this PR and resubmit when it is under 2hr if it is just the time limit blocking this.

(Some time can be shaved off if I didn't need to install Visual Studio Build Tools and Git when they are not already present.)

**Comment by denelon** [comment]
@jonahbeckford the long install time is causing us to time out. The client does have an experimental feature for dependencies, but we would still need to install those for validation. We're looking into options to see if there is something we might be able to do for packages that take an exceptionally long time.

**Comment by jonahbeckford** [comment]
No worries. Thanks for looking. I'll withdraw this PR and resubmit when the install time is shrunk. That will be a win for everyone concerned.

**Pull request closed without merging**